pep795.md

```diff
@@ -0,0 +1,1938 @@
+  A-- gcd -->DD[Function 1]
+  A-- print_fraction -->EE[Function 2]
+  AA-- gcd -->DD
+  BB-- Fraction -->E
+  int -- superclass -->O[object]:::imm
+  E -- superclass -->O
+  O -- \_\_class__ -->type
+
+  B:::imm
+  1:::imm
+  3:::imm
+  int:::imm
+  type:::imm
+  E:::imm
+  AA:::imm
+  BB:::imm
+  CC:::imm
+  DD:::imm
+  classDef imm fill:#ADD8E6,stroke:#00f,stroke-width:1px;
+  linkStyle 1,2,6,8,9,12,13,14,15,16,19,20,21,22,23 stroke:#00f,stroke-width:1px,color:blue;
+```
+
+Note that "for `freeze(f1)` to succeed" is not necessarily
+the same as "these objects are made immutable by a call
+to `freeze(f1)`". Later in this document, we will describe
+our proposal for how freezing can propagate.
+
+The figure above shows that the object pointed to by `f1`
+is immutable as are the objects in its fields (`1` and `3`),
+as well as the object representing the `Fraction` class at
+run-time. Making a class immutable means that operations
+that adds, removes, or changes functions in the class will
+subsequently fail, and will also make any state in the
+class immutable. Since the `Fraction` class object has a
+pointer to the supertype object, `type`, `type` also must be
+immutable. Similarly, freezing the numerator and denominator
+`1` and `3` requires the `int` class to be immutable.
+
+When the `Fraction` class is immutable all its function
+objects, such as `__init__`, `__add__`, and `__repr__` must
+be immutable too, or they are no longer callable. Making a
+function object immutable means that the function is not
+permitted to capture mutable state. Of the three functions
+in `Fraction`,  only `__init__` captures external state: a
+reference to `gcd`. This means that the `gcd` function must
+be made immutable too. Notably, `__add__` has a reference
+to `Fraction` which would  also have to be made immutable,
+but is alredy immutable in this example. (It's becoming
+immutable is what causes `__add__` to become immutable in
+the first place.)
+
+Note that the `gcd` variable in the stack frame did not become
+immutable, although it is pointing to an immutable object. It is
+still possible to reassign that variable. Also note that while a
+Python class technically has a reference to all its subclasses, we
+do not freeze subclasses, only superclasses, when a class is
+frozen. We will return to this later and explain when it is
+possible to make exceptions such as this one.
+
+This PEP will make certain Python objects immutable by default.
+This will be discussed in a separate section below. These objects
+include `type`, `object` and `int` as well as all integers like
+`1`, `2`, `3` and `7`. Thus, in the example, 4 objects would
+have been immutable from creation.
+
+
+**In summary:**
+
+* Making an object immutable blocks subsequent mutation of the object
+* Making a class immutable makes its functions and any class state immutable, and blocks changes to the class; functions that could not be made immutable are no longer callable
```

Comment:
This "no longer callable" is ambiguous. Isn't it more the case that they "remain" callable, but will throw an exception on mutation?